add testing cheat sheet and C++ test template

Bai-Li-NOAA opened the pull request:
<!---
Thanks for opening a PR. This commented text will **NOT** appear in the final PR. Toggle between Write and Preview to see what your PR will look like without the comments.
-->

# What is the feature?
* add testing cheat sheet and C++ test template

# How have you implemented the solution?
* add README.md under the /tests directory as a quick reference for
  testing. It includes instructions on creating a test file, preparing
  test data, using helper functions, running tests, and debugging.
* add a C++ test template to inst/templates.
* add a new R function `use_gtest_template()` to create a C++ test file
  using the template and register the test in
  `tests/gtest/CMakeLists.txt`.
* add R unit tests for `use_gtest_template()`.
* add a C++ test `tests/gtest/test_FIMSJson_JsonParser_WriteToFile.cpp`
  to verify error handling instructions outlined in the C++ test template.
* update LogisticSelectivity C++ test to use the new template.
* fix spelling errors.
*
# Does the PR impact any other area of the project, maybe another repo?
* No.

This PR includes changes to the C++ code in `fims_json.hpp` made during debugging, and adds some template code to test built-in error messages. The following files may need a closer review from @msupernaw to ensure everything looks correct:
- `inst/include/utilities/fims_json.hpp`
- `inst/templates/gtest_template.cpp`
- `tests/gtest/test_JasonParse_WriteToFile.cpp`

Bai-Li-NOAA commented on tests/README.md:
```diff
@@ -0,0 +1,104 @@
+continuous integration runs.
+1. Add `testthat::skip_on_covr()` at the beginning of a test file to skip it
+during coverage calculation.
+1. Run `devtools::test()` to execute all tests in the package.
```
- Update line 69: run `devtools::load()` before running `devtools::test()`. 
- Add a bullet point to mention running `devtools::install()` before running `test-parallel-*.R`

iantaylor-NOAA commented:
@Bai-Li-NOAA, thank you for creating this template. Based on your suggestion in the implementation team notes, the length-based selectivity group just tried out `use_gtest_template()` and have the following feedback

- the function worked perfectly but we didn't have enough time to fill in the details to get a working test 
- it would be nice to not require all the arguments to run the function and have placeholder text as default (e.g. header_file_path = "TODO: add path to header file here")
- we tried to access the template via `source("R/use-gtest-template.R")` rather than wait for FIMS to build but this doesn't work because it can't find the templates (not a big deal)

Bai-Li-NOAA commented:
@kellijohnson-NOAA This PR is ready for review!

Bai-Li-NOAA commented:
@kellijohnson-NOAA, I’ve rebased the dev-add-test-documentation branch.

In addition to adding the testing cheat sheet and C++ test template, I also updated `zzz.R`. Let me know if the documentation about placing `Rcpp::loadModule(module = "fims", what = TRUE)` inside the `.onLoad()` function isn’t clear or helpful.

I tested the changes by creating a custom GHA workflow. It runs the following code and passes on all three operating systems (see logs [here](https://github.com/NOAA-FIMS/FIMS/actions/runs/16455422994)):
```r
devtools::load_all()
devtools::test()
devtools::test()
devtools::install()
testthat::test_package("FIMS")
```
This setup mimics both types of workflows: loading FIMS from source and testing, and loading the installed version of FIMS and testing.

kellijohnson-NOAA commented:
@Bai-Li-NOAA thank you for all of this great work. I plan on reviewing it later today.

kellijohnson-NOAA commented on tests/integration/debug-integration-tmb-nointerface.R:
@Bai-Li-NOAA I just want to double check that you meant to remove all three files in this folder?

Bai-Li-NOAA commented:
@kellijohnson-NOAA, I've finished updating `use_gtest_template()` and its associated tests. In addition, the two commits I pushed:
 - add instructions to suppress messages in `tests/README.md`
 - set initial value of `log_q` to 0 in estimation tests to address the comment [here](https://github.com/NOAA-FIMS/FIMS/issues/919#issuecomment-3119952602). 
 
Let me know if there’s anything else that should be updated before I squash the commits.
